functional_programming_tests/test/with_functional_programming.cpp

```diff
@@ -0,0 +1,94 @@
+#include <gtest/gtest.h>
+
+#include <iostream>
+#include <vector>
+#include <cmath>
+#include <optional>
+
+// production_code.h/cc
+template<typename T>
+using Map = std::vector<std::vector<T>>;
+
+// X, Y position
+struct Pos {
+    size_t x;
+    size_t y;
+};
+
+using Path = std::vector<Pos>;
+
+// Operator overload to print path
+std::ostream& operator<<(std::ostream& os, std::vector<Pos> const& path) {
+    for (auto const& pose:path) {
+        os << "(" << pose.x << ", " << pose.y << ")\n";
+    }
+    return os;
+}
+
+// Operator overload for position comparison
+bool operator==(Pos const& lhs, Pos const& rhs) {
+    return lhs.x == rhs.x && lhs.y == rhs.y;
+}
+
+// From the Robot, goal and costmap, generate a trajectory (Deterministic calculation)
+std::optional<Path> generate_global_path(Pos const& start, Pos const& goal , Map<unsigned char> const& costmap) { // Calculation
+    // Some cool and nifty algorithm
+    // What is the delta in position
+    int const del_x = goal.x - start.x;
+    int const del_y = goal.y - start.y;
+
+    // What direction to move in for each dimension
+    int const del_x_sign = std::copysign(1.0,del_x);
+    int const del_y_sign = std::copysign(1.0,del_y);
+
+    // Push start onto the path
+    Path path;
+    path.push_back(start);
+
+    // Fails if there is any obstacle in the way
+    // Move horizontally
+    for (size_t i = 0; i < (std::abs(del_x)); ++i) {
+        if (costmap.at(path.back().y).at(path.back().x + del_x_sign) == 1) {
```

Comment:
For... reasons... I would add a method to your `Map` class like `Map::at(Position const& position)`

Reply:
Also, you might consider writing a little lambda called `is_occupied`

Reply:
Done

Reply:
I think you haven't committed this change?